### PR TITLE
Single source of truth

### PR DESCRIPTION
**Pull request by pjforde1978**
# Type of PR (feature, enhancement, bug fix, etc.)

Enhancement

## Description

Proposed fix for the problem of multiple users clobbering each other by typing into the same input at the same time. Concept is that if a textual input element has focus, morphdom should never overwrite the input's current value even when it deviates from the server's rendered updated.

## Why should this be added

While not a complete solve for multi-user concurrency, this imposes some sanity on an otherwise imperfect status quo.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] Checks (StandardRB & Prettier-Standard) are passing

### CODE REVIEW

**Comment by hopsoft** [comment]
See #77 as it's related. There was a bug that assumed that connection identifiers are always Rails model instances _or at least object with an `id`_.

**Comment by pjforde1978** [comment]
Great news that #77 was patched, although technically this PR was coming regardless. The server should not be able to overwrite the contents of an input element while you're typing into it.

File: javascript/attributes.js [hunk 1]
```diff
@@ -83,3 +83,33 @@ export const findElement = attributes => {
...
+      'number',
+      'password',
+      'search',
+      'tel',
+      'text',
+      'textarea',
+      'url'
+    ].indexOf(element.type) >= 0
+  )
+}
```
Comment (hopsoft) [inline javascript/attributes.js]: I might add an additional check for `element.tagName`

Comment (pjforde1978) [reply javascript/attributes.js]: Interesting. I thought they were basically the same. Hard to Google. Can you explain the difference in your own

Comment (hopsoft) [reply javascript/attributes.js]:
```javascript
document.getElementsByTagName("INPUT")[0].tagName
"INPUT"

document.getElementsByTagName("INPUT")[0].type
"hidden"
```

This feature should ignore all tags but `input` and `textarea`. Then only apply the logic to the types you've already outlined.

File: javascript/attributes.js [hunk 2]
```diff
@@ -83,3 +83,33 @@ export const findElement = attributes => {
...
+  const element = event.target
+  if (!isTextInput(element)) return
+  element.reflexPermanent = element.hasAttribute('data-reflex-permanent')
+  element.setAttribute('data-reflex-permanent', '')
+}
+
+export const lostFocus = event => {
+  const element = event.target
+  if (!isTextInput(element)) return
+  if (element.reflexPermanent !== undefined && !element.reflexPermanent) {
```
Comment (hopsoft) [inline javascript/attributes.js]: I'd change this guard to:

```javascript
if (element.reflexPermanent) return
```

**Comment by hopsoft** [comment]
I'm currently building a chat demo on the expo project and am seeing a usecase where we actually do want the input clobbered when the reflex updates the DOM. 🤔

**Comment by pjforde1978** [comment]
https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.html#module-custom-dom-patching ???

**Comment by pjforde1978** [comment]
Just wanted to check in on this and see where your head is at. I'm happy to make the changes you suggested, but I wanted to make sure that we're still on the same page about it being a good idea.

**Comment by hopsoft** [comment]
I still think this is a good idea. If someone want's to take action on one of these input fields (as my chat example above) they can still do so via a lifecycle method.

File: javascript/stimulus_reflex.js [hunk 3]
```diff
@@ -285,6 +292,8 @@ if (!document.stimulusReflexInitialized) {
     invokeLifecycleMethod('error', target, element)
     invokeLifecycleMethod('after', target, element)
   })
+  document.addEventListener('focus', receivedFocus, true)
```
Comment (pjforde1978) [inline javascript/stimulus_reflex.js]: Only just learned about [focusin](https://developer.mozilla.org/en-US/docs/Web/API/Element/focusin_event) from your PR. I'm going to commit a patch here that switches over...

Comment (pjforde1978) [reply javascript/stimulus_reflex.js]: Love all of the cool things I learn from @hopsoft :heart_eyes_cat:

**Comment by pjforde1978** [comment]
Oh, awesome! Seemed like you were on the fence.

**Pull request merged**